add shader model 3 support and rewrite dx9 initialization functions to add d3d9on12 support

Berthalamew opened the pull request:
- update atmospheric fog apply shader to be partial precision
- reorder render calls in render_scene to draw sky later on (improving performance)
- fix condition in simulation_queue_entity_decode_header

Berthalamew committed:
use pixel shader 3 bytecode for shaders packaged in map files

Berthalamew committed:
rewrite render_scene

- remove a stretch rect exclusive to shader model 3 code if we're on d3d9ex. we still keep this check if we're on stock d3d9
- rewrite rasterizer_settings_set_antialiasing to fix buffer passed to the CheckDeviceMultiSampleType function

Berthalamew committed:
adjust includes

Berthalamew committed:
use partial precision for the fog_atmospheric_apply shader

- update build script to allow specific parameters per shader

Berthalamew committed:
adjust function calls unicode

Berthalamew committed:
misc additions and changes

- add hash table struct
- add game_in_editor function
- add csmemmove
- add network adapter functions
- make certain xlive functions public
- fix condition in simulation_queue_entity_decode_header
- run tag fixes in the mainmenu

Berthalamew committed:
rewrite dx9 initialization functions to fix SM3 and d3d9ex support

- fixup render_scene to render sky later on and render the texaccum layers
- add d3d9on12 support (not compiled by defualt)
- encase error logging with in the ERRORS_ENABLED macro

Berthalamew committed:
fix device lost issues for ztarget texture without d3d9ex

- fix rasterizer_dx9_prime_shader_initialize function
-

Berthalamew committed:
fix warnings

- mixing signed vs unsigned comparisons
- make sure values in direction table are floats
- make e_saved_game_file_type a signed enum

Berthalamew committed:
fix dof rendering in SM3

- fixup some parameters passed to rasterizer_dx9_set_render_state functions

nukeulater commented on xlive/Blam/Engine/render/render.cpp:
```diff
@@ -304,6 +348,428 @@ void __cdecl render_window(window_bound* window, bool is_texture_camera)
+            }
+
+            rasterizer_dx9_perf_event_begin("texaccum", NULL);
+            if (prepare_render_layer(_render_layer_texture_accumulate))
+            {
+                draw_render_layer();
+                reset_after_render_layer_draw();
+            }
```
function instead? this sequence is quite repetitive

Berthalamew replied:
added the DRAW_RENDER_LAYER macro

Berthalamew committed:
add DRAW_RENDER_LAYER macro

- rename draw_render_layer to render_scene_geometry
- remove unneeded fog patchcall
-

Berthalamew committed:
fix patchy fog when using SM3

Berthalamew committed:
add options to config to force disable sm3 and d3d9ex

- fixup rendering on dx9 since the rasterizer_pixel_shader_index wasn't being set for d3d9

Berthalamew committed:
set windows 10 SDK to the last version listed in the visual studio installer

- if we keep using 10.0 then it'll grab from the windows 11 SDK (if it's installed on the system) - if we compile on the windows 11 SDK it causes a crash on certain hardware

Berthalamew committed:
remove texaccum rendering

- hired gun removed it for a reason, (graphical issues on outskirts)
- fix z target staging in c_shader_submission_interface_new::stage_texture for shader model 3
- fix assertions in simulation code on debug builds
- fix fog rendering in shader model 2

Berthalamew committed:
fix D3D9On12 usage


Co-authored-by: Haydn Trigg <me@haydntrigg.com>

Berthalamew committed:
cleanup d3d9on12 code and add config option for d3d9on12

Berthalamew committed:
automatically disable sm3 if we're on an AMD or ATI gpu

- add s_rasterizer_dx9_driver_globals structure
- add config options to disable the amd or ati patches we do (sm3 diable)

Berthalamew committed:
misc changes

- remove tag callbacks
- use bitmask for g_draw_hud_user_draw_player_indicators since c_static_flags_no_init errors with a k_maximum_bit_count value less than 32
- fix warnings for implicit casting
- remove redundant functions from ObserverMode
- remove redundant if check in saved_games_get_file_info
- remove some xlive code that's never called
- fix empty string being duplicated in ComVar
